[CALCITE-6805] Support hex function for Hive and Spark Library

xuzifu666 opened the pull request:
Currently Spark Hive hex and unhex function not introduce for calcite, need to support it.

jira: https://issues.apache.org/jira/browse/CALCITE-6805

rubenada commented on core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java:
```diff
@@ -1890,6 +1890,26 @@ private static RelDataType deriveTypeMapFromEntries(SqlOperatorBinding opBinding
+          SqlFunctionCategory.STRING);
+
+  /**
+   * The "HEX(string)" function; converts {@code string} into a hexadecimal varchar.
```
Wrong javadoc description (copy-paste error from previous method)

xuzifu666 replied:
ditto

xuzifu666 commented:
@rubenada Thanks for your review, all comments had addressed, PTAL

rubenada commented on core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:
```diff
@@ -359,6 +359,22 @@ public static ByteString fromHex(String hex) {
     }
   }
 
+  /** SQL HEX(varchar) function. */
```
nit: could you please put the new methods after the existing `toHex` method definition? I'd like to keep "related" methods (fromHex/toHex) right next to each other (similarly to the new ones hex/unhex).

xuzifu666 replied:
Done

rubenada commented:
Thanks @xuzifu666 . I left another minor comment, but overall lgtm.
One small detail: there's a typo on the Jira title, PR title and (most importantly) on the commit message (`Libaray` instead of `Library`), could you please fix that?

xuzifu666 commented:
> Thanks @xuzifu666 . I left another minor comment, but overall lgtm. One small detail: there's a typo on the Jira title, PR title and (most importantly) on the commit message (`Libaray` instead of `Library`), could you please fix that?

Very Thanks for @rubenada review , had addressed comments and fix the title of pr and commit, PTAL~

caicancai commented on testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:
```diff
@@ -5823,6 +5823,38 @@ void testBitGetFunc(SqlOperatorFixture f, String functionName) {
     f0.forEachLibrary(libraries, consumer);
   }
 
+  @Test void testHex() {
```
Add jira link

xuzifu666 replied:
Had add spark and hive function link in jira.

caicancai commented on testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:
```diff
@@ -5823,6 +5823,38 @@ void testBitGetFunc(SqlOperatorFixture f, String functionName) {
+          "abc",
+          "VARCHAR NOT NULL");
+      f.checkString("unhex('')", "", "VARCHAR NOT NULL");
+      f.checkNull("unhex(cast(null as varbinary))");
```
You need to add more tests

e.g. 
```
scala> val df = spark.sql("select unhex(null)")
df: org.apache.spark.sql.DataFrame = [unhex(NULL): binary]

scala> df.show()
+-----------+
|unhex(NULL)|
+-----------+
|       null|
+-----------+


scala> val df = spark.sql("select unhex(0)")
df: org.apache.spark.sql.DataFrame = [unhex(0): binary]

scala> df.show()
+--------+
|unhex(0)|
+--------+
|    [00]|
+--------+


scala> val df = spark.sql("select unhex(-1)")
df: org.apache.spark.sql.DataFrame = [unhex(-1): binary]

scala> df.show()
+---------+
|unhex(-1)|
+---------+
|     null|
+---------+


```

xuzifu666 replied:
done

caicancai replied:
`unhex(0) == [00]` not null

xuzifu666 replied:
Yes, unhex maybe had some problem in hive and spark side, after discussion I remove the implement of unhex from this pr, this pr change to support hex only.

rubenada replied:
@xuzifu666 feel free to open a separate Jira ticket (related to this one) to cover the `unhex` implementation, even if you don't plan to work on it right away.

caicancai commented on testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:
```diff
@@ -5823,6 +5823,38 @@ void testBitGetFunc(SqlOperatorFixture f, String functionName) {
+          "616263",
+          "VARCHAR NOT NULL");
+      f.checkString("hex('')", "", "VARCHAR NOT NULL");
+      f.checkNull("hex(cast(null as varbinary))");
```
above all

xuzifu666 replied:
done

caicancai replied:
You need to add a test for numbers

```
scala> val df = spark.sql("select hex('1')")
df: org.apache.spark.sql.DataFrame = [hex(1): string]

scala> df.show()
+------+                                                                        
|hex(1)|
+------+
|    31|
+------+


scala> val df = spark.sql("select hex('0')")
df: org.apache.spark.sql.DataFrame = [hex(0): string]

scala> df.show()
+------+
|hex(0)|
+------+
|    30|
+------+


scala> val df = spark.sql("select hex('-1')")
df: org.apache.spark.sql.DataFrame = [hex(-1): string]

scala> df.show()
+-------+
|hex(-1)|
+-------+
|   2D31|
+-------+
```

xuzifu666 commented:
@caicancai had addressed comments，PTAL

caicancai commented on core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:
```diff
@@ -364,6 +364,21 @@ public static String toHex(ByteString byteString) {
 
+  /** SQL HEX(varchar) function. */
+  public static String hex(String value) {
+    return Hex.encodeHexString(value.getBytes(UTF_8));
```
I'm not sure if it's UTF-8, but I'd accept it if it's compatible with all tests

xuzifu666 replied:
Thanks, UTF-8 seems ok for all, I had consider other funciton such md5/base32.. function also use UTF8 and CI is OK.

caicancai reviewed (changes requested):
Need to add some tests

xuzifu666 commented:
> Need to add some tests

Thanks for reminder, I had add more tests for it~ @caicancai

caicancai reviewed (approved):
LGTM

caicancai commented:
Just a small suggestion, you don't need to compress and submit every time, just compress and submit after the final approval.
